### PR TITLE
Prerendering overhaul

### PR DESCRIPTION
**Pull request by Rich-Harris**
# Migration

## General

`+server` pages can be prerendered, too now, by adding the `export const prerender = ..` option to them. By default they are not prerendered, but any `+server` page that is referenced from a prerendered page inherits the `prerender` option of that page unless explicitly set otherwise. This might lead to a situation where you now get a "Cannot prerender a +server file with POST, PATCH, PUT, or DELETE" error. To fix it, add `export const prerender = false;` to the `+server` file where it happens, or move the mutative methods in a different file.

## adapter-static

Instead of only checking the default for prerendering, the actual found routes are now checked for prerendering. If one of them is not prerenderable, an error is thrown unless you set the `fallback` option. If you get this error now, adjust your code so that 
- either the non-prerenderable routes are not crawled anymore
- or the non-prerenderable routes are made prerenderable
- or set the `fallback` option

# PR description

#6356.

* [x] differentiate between prerenderable and non-prerenderable endpoints
* [x] exclude routes with `prerender = true` from generated manifests
* [x] allow `'auto'` (prerender, but also include in manifests) as a value, alongside `true` (only prerender, exclude from manifests) and `false` (never prerender, include in manifests)
* [x] throw in adapter-static if any routes are `prerender = false` or `prerender = 'auto'` (or unknown, in the case of endpoints that are never fetched during prerendering)
* [x] docs 
* [x] idk, probably some tests and stuff

Not in scope for this PR: removing `config.kit.prerender.default` and using defaults from `+layout[.server].js` files. That can be added to #6197 once this is merged.

### Please don't delete this checklist! Before submitting the PR, please make sure you do the following:
- [x] It's really useful if your PR references an issue where it is discussed ahead of time. In many cases, features are absent for a reason. For large changes, please create an RFC: https://github.com/sveltejs/rfcs
- [x] This message body should clearly illustrate what problems it solves.
- [x] Ideally, include a test that fails without this PR but passes with it.

### Tests
- [x] Run the tests with `pnpm test` and lint the project with `pnpm lint` and `pnpm check`

### Changesets
- [x] If your PR makes a change that should be noted in one or more packages' changelogs, generate a changeset by running `pnpm changeset` and following the prompts. All changesets should be `patch` until SvelteKit 1.0

### CODE REVIEW

File: documentation/docs/12-page-options.md [hunk 1]
```diff
@@ -32,22 +32,29 @@ export const hydrate = false;
 
 ### prerender
 
-It's likely that at least some pages of your app can be represented as a simple HTML file generated at build time. These pages can be [_prerendered_](/docs/appendix#prerendering).
+It's likely that at least some routes of your app can be represented as a simple HTML file generated at build time. These routes can be [_prerendered_](/docs/appendix#prerendering).
 
-Prerendering happens automatically for any page with the `prerender` annotation:
+Prerendering happens automatically for any `+page` or `+server` file with the `prerender` annotation:
```
Comment (benmccann) [inline documentation/docs/12-page-options.md:37]: should we add `JS` here to clarify that you can't add it to `+page.svelte`?
```suggestion
Prerendering happens automatically for any `+page` or `+server` JS file with the `prerender` annotation:
```

File: documentation/docs/12-page-options.md [hunk 2]
```diff
@@ -32,22 +32,29 @@ export const hydrate = false;
...
-/// file: +page.js/+page.server.js
+/// file: +page.js/+page.server.js/+server.js
 export const prerender = false;
 ```
 
+Routes with `prerender = true` will be excluded from manifests used for dynamic SSR, making your server (or serverless/edge functions) smaller. In some cases you might want to prerender a route but also include it in the manifest (for example, you want to prerender your most recent/popular content but server-render the long tail) — for these cases, there's a third option, 'auto':
+
+```js
+/// file: +page.js/+page.server.js/+server.js
+export const prerender = 'auto';
```
Comment (benmccann) [inline documentation/docs/12-page-options.md:55]: now that I've seen the docs and understand what this option does, I'll throw another name into the ring. how about `'partial'`?

File: documentation/docs/12-page-options.md (hunk 1 above)
Comment (Rich-Harris) [reply documentation/docs/12-page-options.md:37]: I was hoping the code sample would make it clear — adding JS muddies the waters a bit in the case where people are using TypeScript

File: documentation/docs/12-page-options.md (hunk 2 above)
Comment (Rich-Harris) [reply documentation/docs/12-page-options.md:55]: I think I prefer `auto` — `partial` doesn't have the right connotation to me, it sounds like the page is being partially prerendered

File: documentation/docs/12-page-options.md (hunk 1 above)
Comment (dummdidumm) [reply documentation/docs/12-page-options.md:37]: There are some other places where it's `+page/layout` with the `.js` ending. We should probably make that consistent, either adding the `.js` file always or never (except for when we introduce it).

File: packages/kit/types/private.d.ts [hunk 3]
```diff
@@ -184,6 +184,10 @@ export interface PrerenderErrorHandler {
 
 export type PrerenderOnErrorValue = 'fail' | 'continue' | PrerenderErrorHandler;
 
+export type PrerenderOption = boolean | 'auto';
```
Comment (dummdidumm) [inline packages/kit/types/private.d.ts:187]: This should be public API, people might want to do `export let prerender: PrerenderOption = ...` or `myHelperDefiningWhatPrerenderOptionToUse(): PreprenderOption {..}`

File: packages/kit/src/runtime/server/page/fetch.js [hunk 4]
```diff
@@ -10,9 +10,10 @@ import { domain_matches, path_matches } from './cookie.js';
  *   options: import('types').SSROptions;
  *   state: import('types').SSRState;
  *   route: import('types').SSRRoute | import('types').SSRErrorPage;
+ *   prerender_default?: true | false | 'auto';
```
Comment (dummdidumm) [inline packages/kit/src/runtime/server/page/fetch.js:13]: This can reuse the PrerenderOption type

**Review by dummdidumm — APPROVED** [review]
Looks good! Will adjust the minor suggestions I made myself and merge to get the related PR in good state

File: documentation/docs/12-page-options.md [hunk 5]
```diff
@@ -32,22 +32,29 @@ export const hydrate = false;
...
 
-It's likely that at least some pages of your app can be represented as a simple HTML file generated at build time. These pages can be [_prerendered_](/docs/appendix#prerendering).
+It's likely that at least some routes of your app can be represented as a simple HTML file generated at build time. These routes can be [_prerendered_](/docs/appendix#prerendering).
 
-Prerendering happens automatically for any page with the `prerender` annotation:
+Prerendering happens automatically for any `+page` or `+server` file with the `prerender` annotation:
 
 ```js
-/// file: +page.js/+page.server.js
+/// file: +page.js/+page.server.js/+server.js
```
[the remark this answers is not in the document]
Comment (dummdidumm) [reply documentation/docs/12-page-options.md:40]: Skipping this for now, seing that it's like this for all others as well.

File: documentation/docs/12-page-options.md [hunk 6]
```diff
@@ -32,22 +32,29 @@ export const hydrate = false;
...
 
 Alternatively, you can set [`config.kit.prerender.default`](/docs/configuration#prerender) to `true` and prerender everything except pages that are explicitly marked as _not_ prerenderable:
 
 ```js
-/// file: +page.js/+page.server.js
+/// file: +page.js/+page.server.js/+server.js
 export const prerender = false;
 ```
 
+Routes with `prerender = true` will be excluded from manifests used for dynamic SSR, making your server (or serverless/edge functions) smaller. In some cases you might want to prerender a route but also include it in the manifest (for example, you want to prerender your most recent/popular content but server-render the long tail) — for these cases, there's a third option, 'auto':
```
Comment (dummdidumm) [inline documentation/docs/12-page-options.md:51]: It took me a little while to understand why this works - why can I prerender AND SSR a route - until I thought "oh yes with `[slug]`". I don't know how to express this better, but maybe we can come up with something later.

**Pull request merged**